src/Utils/LexicalDistance.php

```diff
@@ -0,0 +1,133 @@
+<?php declare(strict_types=1);
+
+namespace GraphQL\Utils;
+
+/**
+ * Computes the lexical distance between strings A and B.
+ *
+ * The "distance" between two strings is given by counting the minimum number
+ * of edits needed to transform string A into string B. An edit can be an
+ * insertion, deletion, or substitution of a single character, or a swap of two
+ * adjacent characters.
+ *
+ * Includes a custom alteration from Damerau-Levenshtein to treat case changes
+ * as a single edit which helps identify mis-cased values with an edit distance
+ * of 1.
+ *
+ * This distance can be useful for detecting typos in input or sorting
+ */
+class LexicalDistance
+{
+    private string $input;
+
+    private string $inputLowerCase;
+
+    /**
+     * @var array<int>
+     */
+    private array $inputArray;
+
+    /**
+     * @var array<array<int>>
+     */
+    private array $rows;
+
+    public function __construct(string $input)
+    {
+        $this->input = $input;
+        $this->inputLowerCase = \strtolower($input);
+        $this->inputArray = self::stringToArray($this->inputLowerCase);
+
+        $length = \mb_strlen($input);
+        $this->rows = [
+            \array_fill(0, $length, 0),
+            \array_fill(0, $length, 0),
+            \array_fill(0, $length, 0),
+        ];
+    }
+
+    public function measure(string $option, float $threshold): ?int
+    {
+        if ($this->input === $option) {
+            return 0;
+        }
+
+        $optionLowerCase = \strtolower($option);
+
+        // Any case change counts as a single edit
+        if ($this->inputLowerCase === $optionLowerCase) {
+            return 1;
+        }
+
+        $a = self::stringToArray($optionLowerCase);
+        $b = $this->inputArray;
+
+        if (\count($a) < \count($b)) {
+            $tmp = $a;
+            $a = $b;
+            $b = $tmp;
+        }
+
+        $aLength = \count($a);
+        $bLength = \count($b);
+
+        if ($aLength - $bLength > $threshold) {
+            return null;
```

Comment:
What does this mean?

Reply:
An early return, optimisation: https://github.com/graphql/graphql-js/blob/v16.3.0/src/jsutils/suggestionList.ts#L84

`levenshtein` function always returns `int`, whereas `LexicalDistance::measure()` returns `null|int`. In contrast with `levenshtein()`, `measure()` takes into account the threshold and returns `null` if the distance is bigger (see changes to `Utils::suggestionList()`, this logic was previously there).

Reply:
Thanks for the explanation, I would like to see that in a comment.